tests/nc-client/main.c

```diff
@@ -0,0 +1,175 @@
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+
+#include "events.h"
+#include "network.h"
+#include "sock.h"
+#include "warnp.h"
+
+struct senddata {
+	char * buffer;
+	ssize_t nchars;
+	int socket;
+	int conndone;
+	void * connect_cookie;
+	void * write_cookie;
+};
+
+/* Forward declarations. */
+static int send_input(void *);
+static int callback_wrote(void * cookie, ssize_t lenwrit);
+static int callback_connected(void * cookie, int socket);
+
+/* Send data from stdin to a socket, or close the connection. */
+static int
+send_input(void * cookie)
+{
+	struct senddata * send = (struct senddata *)cookie;
+	size_t len = 0;
+
+	/* Read data from stdin. */
+	if ((send->nchars = getline(&send->buffer, &len, stdin)) != -1) {
+		/* Send data to server. */
+		if ((send->write_cookie = network_write(send->socket,
+		    (uint8_t *)send->buffer, (size_t)send->nchars,
+		    (size_t)send->nchars, callback_wrote, cookie)) == NULL) {
+			warn0("network_write failure");
+			goto err0;
+		}
+	} else {
+		/* Close connection without sending anything. */
+		send->conndone = 1;
+	}
+
+	/* Success! */
+	return (0);
+
+err0:
+	/* Failure! */
+	return (1);
+}
+
+/* Finished writing data; look for more from stdin. */
+static int
+callback_wrote(void * cookie, ssize_t lenwrit)
+{
+	struct senddata * send = (struct senddata *)cookie;
+
+	/* We are not longer writing. */
+	send->write_cookie = NULL;
+
+	/* Check that we wrote all our data. */
+	if (lenwrit != send->nchars) {
+		warn0("Mismatch in data sent and data requested to send");
+		goto err0;
+	}
+
+	/* Clear the buffer that was used. */
+	free(send->buffer);
+	send->buffer = NULL;
+
+	return send_input(cookie);
+
+err0:
+	/* Failure! */
+	return (1);
+}
+
+/* Got a connection; look for data from stdin. */
+static int
+callback_connected(void * cookie, int socket)
+{
+	struct senddata * send = (struct senddata *)cookie;
+
+	/* We are not longer connecting. */
+	send->connect_cookie = NULL;
+
+	/* Check that the connection did not fail. */
+	if (socket == -1) {
+		warn0("failed to connect");
+		goto err0;
+	}
+
+	/* Record socket for future use. */
+	send->socket = socket;
+
+	return send_input(cookie);
+
+err0:
+	/* Failure! */
+	return (1);
```

Comment:
Whoops, I was thinking about error codes.  Should the `return (1)` in `proto_crypt_dh_validate()` be changed?